book/src/dev/rfcs/0002-parallel-verification.md

```diff
@@ -0,0 +1,335 @@
+                  the genesis block.
+* **genesis:** The root of the tree of blocks is called the genesis block. It has
+               no previous block.
+* **tip:** A block which has no next block is called a tip. Each chain fork can
+           be identified using its tip.
+
+Data:
+* **consensus rule:** A protocol rule which all nodes must apply consistently,
+                      so they can converge on the same chain fork.
+* **context-free:** Consensus rules which do not have a data dependency on
+                    previous blocks.
+* **data dependency:** Information contained in the previous block and its
+                       chain fork, which is required to verify the current block.
+* **semantic verification:** Verifying the consensus rules on the data structures
+                             defined by the protocol.
+* **state:** The set of verified blocks. The state may also cache some dependent
+             data, so that we can efficienty verify subsequent blocks.
+* **structural verification:** Parsing raw bytes into the data structures defined
+                               by the protocol.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In Zebra, we want to verify blocks in parallel. Some fields can be verified
+straight away, because they don't depend on the output of previous blocks.
+But other fields have **data dependencies**, which means that we need previous
+blocks before we can fully validate them.
+
+If we delay checking some of these data dependencies, then we can do more of
+the verification in parallel.
+
+## Example: BlockHeight
+[block-height]: #block-height
+
+Here's how Zebra can verify the different `BlockHeight` consensus rules in
+parallel:
+
+**Parsing:**
+
+1. Parse the Block into a BlockHeader and a list of transactions.
+
+**Verification - No Data Dependencies:**
+
+2. Check that all BlockHeights are within the range of valid heights.^
+3. Check that the block has exactly one BlockHeight.
+4. Check that the BlockHeight is in the first transaction in the Block.
+
+**Verification - Deferring A Data Dependency:**
+
+5. Verify other consensus rules that depend on BlockHeight, assuming that the
+   BlockHeight is correct. For example, many consensus rules depend on the
+   current Network Upgrade, which is determined by the BlockHeight. We verify
+   these consensus rules, assuming the BlockHeight and Network Upgrade are
+   correct.
+
+**Verification - Checking A Data Dependency:**
+
+6. Await the previous block. When it arrives, check that the BlockHeight of this
+   Block is one more than the BlockHeight of the previous block. If the check
+   passes, commit the block to the state. Otherwise, reject the block as invalid.
+
+^ Note that Zebra actually checks the BlockHeight range during parsing. The
+  BlockHeight is stored as a compact integer, so out-of-range BlockHeights take
+  up additional raw bytes during parsing.
```

Comment:
I don't think this is accurate, the block height is not checked during (block) parsing, and the block height isn't stored as a compact integer, it's stored in script encoding.

Reply:
I'll just delete that then.

Reply:
Fixed in 626065adf65e7a3df3be8fbcb1f34399dd7d044c.